### ISSUE TITLE
`slice.unique` still returns incorrect slice

### ISSUE THREAD

**Issue by skaman** [odin-lang/Odin#3769]
## Context

Odin:    dev-2024-06-nightly:f745a1c47
OS:      Windows 11 Professional (version: 23H2), build 22631.3737
CPU:     AMD Ryzen Threadripper 3970X 32-Core Processor
RAM:     65414 MiB
Backend: LLVM 17.0.1

## Expected Behavior

Given this example code

```odin
tmp := slice.unique([]int{1, 2, 4, 4, 5})
for v, i in tmp {
    fmt.printfln("index: %d, value: %d", i, v)
}
```

It should output:

```txt
index: 0, value: 1
index: 1, value: 2
index: 2, value: 4
index: 3, value: 5
```

## Current Behavior

The output that currently i get is:

```txt
index: 0, value: 1
index: 1, value: 4
index: 2, value: 5
```

## Failure Information (for bugs)

I think the bug is related to:
- Issue: https://github.com/odin-lang/Odin/issues/3276
- PR: https://github.com/odin-lang/Odin/pull/3277

In detail the index `i` not increase when it's equal with `j` but it should.

### Steps to Reproduce

1. Paste the example code from "Expected Behavior" into the main proc
2. Run it

**Comment by Kelimion** [comment]
Interesting. The previous fix did indeed fix the examples given there:
```cpp
	s := []int{2,2,2}
	res := slice.unique(s)
	assert(slice.equal(res,[]int{2}))

	res = slice.unique([]int{1,1,1,2,2,3,3,3,3})
	assert(slice.equal(res, []int{1,2,3}))

	res = slice.unique_proc([]int{1,1,1,2,2,3,3,3,3}, proc(a, b: int) -> bool {
		return a == b
	})
	assert(slice.equal(res, []int{1,2,3}))
```
But I can also reproduce the bug you found using that old fix.
I'll have a look at your fix momentarily and will write more extensive test coverage for `tests/core/slice`.

**Comment by Kelimion** [comment]
Fixed by #3770.
